### PR TITLE
README.md: assorted improvements

### PR DESCRIPTION
**Pull request by kolyshkin**
* typos
* @users linked
* formatting fixes (lists, paragraphs, headers etc).
* something else I already forgot

### CODE REVIEW

**Comment by BlackBrix** [comment]
removing typos and increase the readability is a good thing but IMHO the whole readme is somehow outdated,
e.g.:
- I can not confirm that it compiles with 4.9.x, I only compiled & tested successfully with gcc-arm-none-eabi Versions: 4.8, 5.4, 6.0, 7.2 (so: newer versions work as well, but not all ...)
-  you don't need to define the USER_CODE_RAM and RAM_END stuff at all ...
- it should be explained in a better way in which cases you need to define the USB_DISC and the BUTTON defines and in which cases they can be omitted (e.g. when commenting out the HAS_MAPLE_HARDWARE statement)
- in my eyes some statements regarding the buttons are confusing and/or misleading:

> Note. Most "generic" STM32F103 boards only have a reset button, and not a user / test button. So the bootloader code always configures the Button input pin as PullDown, hence if a button is not present on the Button pin (Default is PC14), the pin should remain in a LOW state, and the bootloader will assume that the Button is not being pressed.

> IMPORTANT. If you have a board where you have external hardware attached to pin **PC12** which will pull this pin HIGH, you will need to make a new build target for you board which uses a different pin for the Button, or completely modify the code so that the Button is ignored.

- what buttons are you talking about the reset- or the user button ?
- in the code I can not see any reference to **PC12** - so it is unclear what is really meant here ?
- so if I omit the complete BUTTON_BANK GPIOB -stuff neither PC14 nor PC12 is 'touched' (e.g. configred as pulldown) by the bootloader-code ?

**Comment by kolyshkin** [comment]
@BlackBrix thanks for reviewing this! Indeed you are right, and I have acted upon each item, resulting in a rewrite. Please see the updated patch. A few comments:

> in the code I can not see any reference to PC12 - so it is unclear what is really meant here?

It was a typo (in the original document) -- the author meant PC14. It is referred to in the code as as combination of `BUTTON_BANK` (set to `GPIOC` for generic boards) and `BUTTON_PIN` (set to `14`).

> so if I omit the complete BUTTON_BANK GPIOB -stuff neither PC14 nor PC12 is 'touched' (e.g. configred as pulldown) by the bootloader-code ?

From what I see in the code, looks like yes -- there are `#ifdef` statements in the button code, so in case `BUTTON_*` defines are undefined, nothing is done wrt button. I have modified README to state that.

**Comment by BlackBrix** [comment]
to me it looks very good now ! *thumbsup*

**Comment by kolyshkin** [comment]
@rogerclarkmelbourne PTAL

**Comment by BlackBrix** [comment]
just 2 additional points that came to my mind today:
- explain what are the differences between "binaries" and "bootloader_only_binaries" files (and why this two different variants have been made)
- maybe set a link to the WIKI as well --> https://github.com/rogerclarkmelbourne/Arduino_STM32/wiki/stm32duino-bootloader

**Comment by rogerclarkmelbourne** [comment]
See http://www.stm32duino.com/viewtopic.php?t=2376  it explains why I had to make binaries with a built in sketch, as most newbie Arduino users were confused that the Bootloader does not appear as a serial device when they first installed it

For legacy linking issues, I had to use the "binaries" folder to contain the combined sketch and bootloader.

This is not ideal but it saves loads posts to here and to the forum, where people want the combined bootloader and dummy sketch and install the raw bootloader.

**Comment by kolyshkin** [comment]
> just 2 additional points that came to my mind today

Makes sense, but it can be added later.

File: README.md [hunk 1]
```diff
@@ -1,101 +1,95 @@
...
 
-e.g.
+Most "generic" STM32F103 boards only have one button -- a reset one, and lack a second user/test button. The second button is also used by the bootloader to enter the "permanent bootloader" state, and for generic boards its pin is set to `PC14` in config.h. The bootloader code configures a the button input pin as pull-down, hence if a button is not physically present on that pin (`PC14`), the pin should remain in a LOW state, and the bootloader will assume that the button is not being pressed.
 
-```
+**IMPORTANT:** if you have a board where you have external hardware attached to pin `PC14` which will pull this pin HIGH, you will need to make a new build target for you board to either use a different pin for the button, or omit the `BUTTON_*` defines altogether. In the latter case the bootloader won't do anything special about the second button.
+
+## Targets and configuration
+
+The configuration for each build target / board is defined in `config.h`
```
Comment (Rotzbua) [inline README.md]: point at end

File: README.md [hunk 2]
```diff
@@ -1,101 +1,95 @@
...
+2. which pin the "button" is attached to.
 
-The Makefile sets the board name as a #define e.g. #define TARGET_GENERIC_F103_PD2, and config.h contains a block of config defines for each board.
+## Second button support
 
-e.g.
+Most "generic" STM32F103 boards only have one button -- a reset one, and lack a second user/test button. The second button is also used by the bootloader to enter the "permanent bootloader" state, and for generic boards its pin is set to `PC14` in config.h. The bootloader code configures a the button input pin as pull-down, hence if a button is not physically present on that pin (`PC14`), the pin should remain in a LOW state, and the bootloader will assume that the button is not being pressed.
 
-```
+**IMPORTANT:** if you have a board where you have external hardware attached to pin `PC14` which will pull this pin HIGH, you will need to make a new build target for you board to either use a different pin for the button, or omit the `BUTTON_*` defines altogether. In the latter case the bootloader won't do anything special about the second button.
```
Comment (Rotzbua) [inline README.md]: If

File: README.md [hunk 3]
```diff
@@ -1,101 +1,95 @@
...
-If you have questions about the bootloader please raise and issue and I will attempt to answer them.
+## Other improvements on the original Maple bootloader
 
+1. Smaller footprint, now only 8k (this became possible due to code changes by [@jonatanolofsson](https://github.com/jonatanolofsson) which allowed the GCC optimisation for size flag to be used.
 
+2. Additional DFU AltID upload type was added, which allows the sketch to be loaded at 0x8002000 instead of 0x8005000 (due to reduced size of the bootloader itself). **Note** upload to 0x8005000 was retained for backwards compatibility.
 
+## Feedback
 
+If you have questions about the bootloader, please raise an issue and I will attempt to answer it.
```
Comment (Rotzbua) [inline README.md]: for questions the forum? http://www.stm32duino.com
and bugs a new issue?

Comment (kolyshkin) [reply README.md]: I would leave it as is (I am not changing this text in my commit, just moving it around), as it's author's prerogative to specify the ways to contact him.

File: README.md (hunk 1 above)
Comment (kolyshkin) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (kolyshkin) [reply README.md]: fixed

**Comment by florisla** [comment]
This issue can be closed.

**Pull request closed without merging**